janusgraph-hadoop-parent/janusgraph-hadoop-core/src/main/java/org/janusgraph/hadoop/formats/cassandra/CqlBridgeRecordReader.java

```diff
@@ -0,0 +1,372 @@
+import java.util.Map;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * <p> Background: The {@link org.apache.cassandra.hadoop.cql3.CqlRecordReader} class has changed
+ * significantly in Cassandra-3 from Cassandra-2. This class acts as a bridge between
+ * CqlRecordReader in Cassandra-2 to Cassandra-3. In essence, this class recreates CqlRecordReader
+ * from Cassandra-3 without referring to it (because otherwise we'd get functionality from
+ * CqlRecordReader on Cassandra-2 and we don't want it). </p>
+ *
+ * @see <a href="https://github.com/JanusGraph/janusgraph/issues/172">Issue 172</a>.
+ */
+@Unstable // Because it should go away with JanusGraph upgrading to Cassandra-3 for OLAP
+@Deprecated // Because this class should already be on its path to deprecation
+public class CqlBridgeRecordReader extends RecordReader<StaticBuffer, Iterable<Entry>> {
+
+    /* Implementation note: This is inspired by Cassandra-3's org/apache/cassandra/hadoop/cql3/CqlRecordReader.java */
+    private static final Logger log = LoggerFactory.getLogger(CqlBridgeRecordReader.class);
+
+    private ColumnFamilySplit split;
+    private DistinctKeyIterator distinctKeyIterator;
+    private int totalRowCount; // total number of rows to fetch
+    private String keyspace;
+    private String cfName;
+    private String cqlQuery;
+    private Cluster cluster;
+    private Session session;
+    private IPartitioner partitioner;
+    private String inputColumns;
+    private String userDefinedWhereClauses;
+
+    private List<String> partitionKeys = new ArrayList<>();
+
+    // partition keys -- key aliases
+    private LinkedHashMap<String, Boolean> partitionBoundColumns = Maps.newLinkedHashMap();
+    private int nativeProtocolVersion = 1;
+
+    // binary type mapping code from CassandraBinaryRecordReader
+    private KV currentKV;
+
+    CqlBridgeRecordReader() { //package private
+        super();
+    }
+
+    @Override
+    public void initialize(InputSplit split, TaskAttemptContext context) throws IOException {
+        this.split = (ColumnFamilySplit) split;
+        Configuration conf = HadoopCompat.getConfiguration(context);
+        totalRowCount = (this.split.getLength() < Long.MAX_VALUE)
+            ? (int) this.split.getLength()
+            : ConfigHelper.getInputSplitSize(conf);
+        cfName = ConfigHelper.getInputColumnFamily(conf);
+        keyspace = ConfigHelper.getInputKeyspace(conf);
+        partitioner = ConfigHelper.getInputPartitioner(conf);
+        inputColumns = CqlConfigHelper.getInputcolumns(conf);
+        userDefinedWhereClauses = CqlConfigHelper.getInputWhereClauses(conf);
+
+        try {
+            if (cluster != null) {
+                return;
+            }
+            // create a Cluster instance
+            String[] locations = split.getLocations();
+//            cluster = CqlConfigHelper.getInputCluster(locations, conf);
+            // disregard the conf as it brings some unforeseen issues.
```

Comment:
The good news is that we have tested this code with success on some large graphs. I believe the code should be good to go in JG proper mainly because as-is JG is broken with Cassandra-3 on OLAP path and this is only going to improve the situation. I am committed to making further changes as and when we uncover issues pertaining to this.